ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/DbChangelog.hs

```diff
@@ -0,0 +1,430 @@
+immutableTipSlot =
+      getTipSlot
+    . either unDbChangelogState unDbChangelogState
+    . AS.head
+    . changelogImmutableStates
+
+{-------------------------------------------------------------------------------
+  Flushing
+-------------------------------------------------------------------------------}
+
+-- | The flush policy
+data DbChangelogFlushPolicy =
+    -- | Always flush everything older than the immutable tip
+    DbChangelogFlushAllImmutable
+
+flush ::
+     forall l.
+     (GetTip (l EmptyMK), HasLedgerTables l)
+  => DbChangelogFlushPolicy
+  -> DbChangelog l
+  -> (DbChangelog l, DbChangelog l)
+flush DbChangelogFlushAllImmutable dblog =
+      (ldblog, rdblog)
+  where
+    DbChangelog {
+        changelogDiffAnchor
+      , changelogDiffs
+      , changelogImmutableStates
+      , changelogVolatileStates
+      } = dblog
+
+    imm = changelogImmutableStates
+    vol = changelogVolatileStates
+
+    immTip = AS.anchor vol
+
+    -- TODO #2 by point, not by count, so sequences can be ragged
+    splitSeqDiff ::
+         (Ord k, Eq v)
+      => SeqDiffMK k v
+      -> (SeqDiffMK k v, SeqDiffMK k v)
+    splitSeqDiff (SeqDiffMK sq) =
+        bimap SeqDiffMK SeqDiffMK
+      $ splitAt (AS.length imm) sq
+
+    lr :: LedgerTables l (Product2 SeqDiffMK SeqDiffMK)
+    lr = mapLedgerTables (uncurry Pair2 . splitSeqDiff) changelogDiffs
+
+    l = mapLedgerTables (\(Pair2 x _) -> x) lr
+    r = mapLedgerTables (\(Pair2 _ y) -> y) lr
+
+    ldblog = DbChangelog {
+        changelogDiffAnchor
+      , changelogDiffs           = l
+      , changelogImmutableStates = imm
+      , changelogVolatileStates  = AS.Empty immTip
+      }
+
+    rdblog = DbChangelog {
+        changelogDiffAnchor      = getTipSlot (unDbChangelogState immTip)
+      , changelogDiffs           = r
+      , changelogImmutableStates = AS.Empty immTip
+      , changelogVolatileStates  = vol
+      }
+
+-- | Flush the immutable changes to the backing store
+--
+-- The 'Ouroboros.Consensus.Storage.ChainDB.Impl.LgrDB.LgrDb'
+-- 'Ouroboros.Consensus.Storage.ChainDB.Impl.LgrDB.flushLock' write lock must be
+-- held before calling this function.
+--
+-- PRECONDITION: @dblog@ should only contain the diffs for the immutable part
+-- of the changelog. If not, the @slot@ that we flush to the backing store will
+-- not match the actual tip of the diffs that we flush to the backing store.
+--
+-- == Note [Behaviour of flushIntoBackingStore]
+--
+-- Consider a simplified @DbChangelog@ that consists of a diff, an immutable
+-- part, and a volatile part. By /flush/ we mean @'flushIntoBackingStore'@.
+--
+-- >   changelog = diffs, imm, vol
+-- >   flush     = @'OnDisk.flush'@
+--
+-- Consider a changelog shaped as shown below, which spans blocks a, b, c, and d.
+--
+-- >   a     b     c     d
+-- >   |-----|     |-----|
+-- >     imm         vol
+-- >   |-----------------|
+-- >          diffs
+--
+-- Each block @x@ has an accompanying diff @diff(x)@. If we flush this
```

Comment:
What has a diff, the block or the ledger state that corresponds to the application of that block?

Reply:
I would say then "a block __produces__ a diff".